include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,1175 @@
+ * argument @p op and creates a blockdiagonal linear operator with @p m
+ * copies of it.
+ */
+template <unsigned int m,
+          typename Range = BlockVector<double>,
+          typename Domain = Range>
+LinearOperator<Range, Domain>
+blockdiag_linop(const LinearOperator<typename Range::BlockType, typename Domain::BlockType> &op)
+{
+
+  static_assert(m > 0,
+                "a blockdiagonal LinearOperator must consist of at least "
+                "one block");
+
+  LinearOperator<Range, Domain> return_op;
+
+  return_op.reinit_range_vector = [op](Range &v, bool fast)
+  {
+    // Reinitialize the block vector to m blocks:
+    v.reinit(m);
+
+    // And reinitialize every individual block with reinit_range_vectors:
+    for (unsigned int i = 0; i < m; ++i)
+      op.reinit_range_vector(v.block(i), fast);
+
+    v.collect_sizes();
+  };
+
+  return_op.reinit_domain_vector = [op](Domain &v, bool fast)
+  {
+    // Reinitialize the block vector to m blocks:
+    v.reinit(m);
+
+    // And reinitialize every individual block with reinit_domain_vectors:
+    for (unsigned int i = 0; i < m; ++i)
+      op.reinit_domain_vector(v.block(i), fast);
+
+    v.collect_sizes();
+  };
+
+  return_op.vmult = [op](Range &v, const Domain &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.vmult(v.block(i), u.block(i));
+  };
+
+  return_op.vmult_add = [op](Range &v, const Domain &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.vmult_add(v.block(i), u.block(i));
+  };
+
+  return_op.Tvmult = [op](Domain &v, const Range &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.Tvmult(v.block(i), u.block(i));
+  };
+
+  return_op.Tvmult_add = [op](Domain &v, const Range &u)
+  {
+    Assert(v.n_blocks() == m, ExcDimensionMismatch(v.n_blocks(), m));
+    Assert(u.n_blocks() == m, ExcDimensionMismatch(u.n_blocks(), m));
+
+    for (unsigned int i = 0; i < m; ++i)
+      op.Tvmult_add(v.block(i), u.block(i));
+  };
+
+  return return_op;
+}
+
+
+
+namespace internal
+{
+  /*
+   * A factory class that is responsible for the creation of a
+   * reinit_range_vector object for a given pair of vector type Range and matrix
+   * type Matrix.
+   *
+   * The generic version of this class just calls
+   * <code>Range::reinit()</code> with the result of
+   * <code>Matrix::m()</code>. This class is specialized for more complicated
+   * data structures, such as TrilinosWrappers::MPI::Vector, etc.
+   */
+  template<typename Range>
+  class ReinitRangeFactory
+  {
+  public:
```

Comment:
Could we move this into a namespace  internal::LinearOperator so that it nicely groups in the documentation?